blog/2023-08-21-soc-vertx-pinot-client.mdx

```diff
@@ -0,0 +1,135 @@
+to monitors their orders in realtime. The dashboard deals with three types of data: products, users, and orders. The
+dashboard uses the Vert.x Pinot Client to [`query`](https://github.com/lucifer4j/pizza-shop-demo/blob/main/dashboard/backend/src/main/java/pizzashop/DashboardVerticle.java)
+the Pinot cluster.
+
+![Order Dashboard Statistics](/images/blog/soc-vertx-pinot-client/demo-dashboard-1.png)
+![Order Dashboard Latest and Popular Items](/images/blog/soc-vertx-pinot-client/demo-dashboard-2.png)
+
+## Challenges
+
+I worked on this project as a part of the [Google Summer of Code](summerofcode.withgoogle.com/) - 2023 program and
+during the implementation, I encountered several challenges.
+
+#### Package-private `org.apache.pinot.client.BrokerResponse` class
+
+The [`org.apache.pinot.client.BrokerResponse`](https://github.com/apache/pinot/blob/6e235a4ec2a16006337da04e118a435b5bb8f6d8/pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/BrokerResponse.java#L27)
+class is package-private. However several [`executeQuery/executeQueryAsync`](https://github.com/apache/pinot/blob/6e235a4ec2a16006337da04e118a435b5bb8f6d8/pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/PinotClientTransport.java#L29)
+methods in the `PinotClientTransport` return this type. This makes it difficult to implement the interface in an external package.
+
+To work around this issue, we created a `org.apache.pinot.client` package inside our wrapper project. The [`VertxPinotClientTransport`](https://github.com/lucifer4j/pinot-client/blob/70dc4152cc77fd6ae521f94fb8c59bf4c7f9d12a/pinot-client/src/main/java/org/apache/pinot/client/VertxPinotClientTransport.java#L37)
+implementation currently lives inside this package for the same reason. This workaround only works for projects running in `classpath`
+mode but not for projects using `modulepath` because Java Modules disallow package splitting. To fix the issue as source,
+I opened a [`PR`](https://github.com/apache/pinot/pull/10932) in `apache/pinot` repository. The PR is now merged. Once a new
+release of Pinot is available, I will be updating the client wrapper as well.
+
+#### `java.concurrent.util.Future` return type
+Some methods of the `PinotClientTransport` [interface](https://github.com/apache/pinot/blob/6e235a4ec2a16006337da04e118a435b5bb8f6d8/pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/PinotClientTransport.java#L32)
+return a `java.concurrent.util.Future` future. However, `Future.get` is a blocking call which is undesirable. To work around the
+issue, we wrap every request and query call in an [`executeBlocking` block](https://github.com/reactiverse/pinot-client/blob/main/pinot-client/src/main/java/io/reactiverse/pinot/client/impl/Utils.java#L24-L34).
+To avoid blocking calls together, we discussed replacing `Future` with `CompletabeFutre`. The same change has already been [merged](https://github.com/apache/pinot/pull/10326)
```

Comment:
```suggestion
To avoid blocking calls altogether, we discussed replacing `Future` with `CompletabeFutre`. The same change has already been [merged](https://github.com/apache/pinot/pull/10326)
```